contracts/src/main/kotlin/com/contractsAndStates/states/GameBoardState.kt

```diff
@@ -66,31 +86,144 @@ class HexTile(val resourceType: String,
 
-        var newMutableListOfRoads = newMutableHexTiles[this.hexTileIndex].roads.map { it }.toMutableList()
-        newMutableListOfRoads.set(sideIndex, roadStateLinearId)
+        val newMutableListOfRoads = newMutableHexTiles[this.hexTileIndex.value].roads.map { it }.toMutableList()
+        newMutableListOfRoads.set(sideIndex.value, roadStateLinearId)
 
-        val reciprocalSideIndex = if (sideIndex + 3 > 5) sideIndex - 3 else sideIndex + 3
-        if (this.sides[sideIndex] != null) {
-            var newMutableReciprocalListOfRoads = newMutableHexTiles[this.sides[sideIndex]!!].roads.map { it }.toMutableList()
-            newMutableReciprocalListOfRoads.set(reciprocalSideIndex, roadStateLinearId)
-            newMutableHexTiles[this.sides[sideIndex]!!].roads = newMutableReciprocalListOfRoads
+        val reciprocalSideIndex = sideIndex.opposite()
+        val neighbor = this.sides.getNeighborOn(sideIndex)
+        if (neighbor != null) {
+            val newMutableReciprocalListOfRoads = newMutableHexTiles[neighbor.value].roads.map { it }.toMutableList()
+            newMutableReciprocalListOfRoads.set(reciprocalSideIndex.value, roadStateLinearId)
+            newMutableHexTiles[neighbor.value].roads = newMutableReciprocalListOfRoads
         }
 
-        newMutableHexTiles[this.hexTileIndex].roads = newMutableListOfRoads
+        newMutableHexTiles[this.hexTileIndex.value].roads = newMutableListOfRoads
         return newMutableHexTiles
     }
 
 }
 
 @CordaSerializable
-data class PortTile(val inputRequired: List<Amount<TokenType>>, val outputRequired: List<Amount<TokenType>>)
+data class PortTile(val inputRequired: List<Amount<TokenType>>, val outputRequired: List<Amount<TokenType>>) {
+
+    init {
+        val inputTypes = inputRequired.map { it.token }
+        require(inputTypes.size == inputTypes.toSet().size) {
+            "There should be no duplicates in the inputRequired list"
+        }
+        require(inputRequired.none { it.quantity == 0L }) {
+            "No inputRequired should have a 0 quantity"
+        }
+        val outputTypes = outputRequired.map { it.token }
+        require(outputTypes.size == outputTypes.toSet().size) {
+            "There should be no duplicates in the outputRequired list"
+        }
+        require(outputRequired.none { it.quantity == 0L }) {
+            "No outputRequired should have a 0 quantity"
+        }
+    }
+
+    companion object {
+        const val PORT_COUNT = 9
+    }
+}
 
 @CordaSerializable
 data class Port(val portTile: PortTile, var accessPoints: List<AccessPoint>)
 
 @CordaSerializable
-data class AccessPoint(val hexTileIndex: Int, val hexTileCoordinate: List<Int>)
\ No newline at end of file
+data class AccessPoint(val hexTileIndex: HexTileIndex, val hexTileCoordinate: List<TileCornerIndex>)
+
+@CordaSerializable
+data class HexTileNeighbors(private val value: MutableList<HexTileIndex?> = MutableList(HexTile.SIDE_COUNT) { null }) {
```

Comment:
@rogersanick I discovered that when a class is deserialised, the initial value passed as `MutableList` is actually not mutable... This is causing a problems as early as `tb.verify(serviceHub)`.
Is there a workaround or should we migrate to `List`?

BTW, the `private val` is a mistake, I am fixing it to `val` in a future commit.

Reply:
Another way to go would be to have 2 types, the serialisable one with `List`, and the not serialisable one with `MutableList`.